### DISCUSSION TITLE
Downcasting of return values

### DISCUSSION THREAD

**Discussion by nyholku** [bytedeco/javacpp#538] [Q&A] [answered]
I've managed to hack together, compile and test POC for opencascade::handle and it seems to function in trivial cases: creating and passing classes and derived classes to OCC native API.

How should I go about going the other way? When the API returns an opencascade object (via opencascade::handle) I would like to downcast the object to the appropriate actual Java class so that instanceof etc work as expected. 

The ancient jCAE SWIG interface uses something like this but I have not figured out how to inject something like this to the JavaCPP  generated code. I'm not even sure if this it is feasible to do this on the Java side ... should this use some kind of adapter code on the C++ side...

```
    private static Object downcastImpl(long ptr, String cName, Class fallback) {
        Constructor ct = constructors.get(cName);
        try {
            if(ct == null) {
                // TODO: may be something about Handle_ prefix
                Class clazz;
                try {
                    clazz = Class.forName("occjava." + cName);
                } catch(ClassNotFoundException ex) {
                    clazz = fallback;
                }
                ct = clazz.getDeclaredConstructor(Long.TYPE, Boolean.TYPE);
                constructors.put(cName, ct);
            }
            return ct.newInstance(ptr, true);

        // TODO: replace with ReflectiveOperationException ?
        } catch(NoSuchMethodException ex) {
            ex.printStackTrace();
        } catch(InstantiationException ex) {
            ex.printStackTrace();
        } catch(IllegalAccessException ex) {
            ex.printStackTrace();
        } catch(InvocationTargetException ex) {
            ex.printStackTrace();
        }
        return null;
    }
```

PS It seems I somehow miss use this Discussions mechanism in github as I cannot mark any of my threads as answered...

**Comment by saudet** [comment, marked as answer]
We can easily do a reinterpret cast from anything to anything else with the constructors that accept a Pointer argument, like this one:
http://bytedeco.org/javacpp/apidocs/org/bytedeco/javacpp/BytePointer.html#BytePointer-org.bytedeco.javacpp.Pointer-
Or do you mean how to have it do that automatically? That's not something JavaCPP can do automatically right now, but you can provide  your own helper methods for that, that's very straightforward. I'm not sure what your question is... Could you elaborate?

> PS It seems I somehow miss use this Discussions mechanism in github as I cannot mark any of my threads as answered...

Don't mind, just do it when you figure it out :)

**Reply by nyholku** [reply]
I will give you an example of what I would like to do:

In OCC class Brep_Tool there is C++ function like this:

```
class BRep_Tool 
{
public:
<snip>
  Standard_EXPORT static Handle(Geom_Surface) Surface (const TopoDS_Face& F);
```

(Ignore the `Standard_EXPOR` and assume `Handle` is a kind of `shared_ptr` which I think I have managed with the `@Adapter` annotation and the `SmartPtrAdapter` template.)

Now I would like to be able use it like this inJava (which can be done with occjava SWIG bindings):

```
Geom_Surface aSurface = BRep_Tool.Surface(aFace);
if (aSurface instanceof Geom_Plane) { 
    Geom_Plane aPlane = (Geom_Plane)aSurface;
    <.do something  with aPlane>
```

In OCC `Geom_Plane` is derived from `Geom_Surface` and I've managed to configure JavaCPP so that when I call an anything that takes a `Geom_Surface` I can pass `Geom_Plane` instance without a cast ie the inheritance hierarchy is ok and the down cast works automatically fine.

Now I need to go the other way round, ie perform the upcast for the return value correctly. Ie in the above example when the `BRep_Tool.Surface()` returns (in C++ side) an instance of `Geom_Plane` it should results in an instance of the corresponding Java class in the Java code and not an instance of `Geom_Surface`, which is what I get now with my above mentioned adapter approach.

Obviously this needs to rely on RTTI which in OCC is available with `DynamicType()` call for any instance of any class derived from `Standard_Transient`. From the `DynamicType()` call the name of the actual C++ class is available and this should be used (AFAIU) to instantiate a corresponding JavaCPP generated class together with return value from the `BRep_Tool.Surface()` call.

As all JavaCPP generated methods seem to be 'native' I think this needs to handled in C++ side. Some annotation for the return type and some kind of adapter template picked up with template. Or something like this. Creating an instance of Java class dynamically in C++ is more of hassle than on the Java side but should be doable. A Java side solution would be preferable.

**Reply by saudet** [reply]
Ok, so you're talking about RTTI, there's an issue opened about that, see https://github.com/bytedeco/javacpp/issues/419#issuecomment-677940450. We can most likely figure out something with standard C++, but if OCC has featues to make this easier like that `DynamicType()` method, I would just recommend to simply use that. We can do that all in Java, there's probably no need to do anything special in C++, no. It should be pretty straightforward actually. Could you explain where you're having issues with what more concretely? Like an example that doesn't work would probably help clear up any misunderstandings.

**Reply by nyholku** [reply]
I have hard time getting up a better explanation of my problem.

The problem is that I cannot even make a start at this issue.

How do I tell JavaCPP that the return value of native function needs different treatment that a parameter of same type?

Atm in POC I have:
```
	@Documented
	@Retention(RetentionPolicy.RUNTIME)
	@Target({ElementType.METHOD, ElementType.PARAMETER})
	@Adapter("MySmartPtrAdapter") // Refers to the C++ adapter class
	public @interface MySmartPtr { // Creates a Java annotation type to be used for smart pointers
	    /** template type */
	    String value() default "";
	}
```
and
```
    	infoMap.put(new Info("opencascade::handle").skip().annotations("@MySmartPtr"));
```

this generates Java code like:

```
 public native @MySmartPtr Base doMore(@MySmartPtr Base p);
```

from my mock up library code of:
```
opencascade::handle<Derived> someFoo() ;

class Derived : public Base {
	public: opencascade::handle<Base>  doMore(opencascade::handle<Base> p){
		return someFoo();

```

On the C++ side the generated code looks like:

```
JNIEXPORT jobject JNICALL Java_occ_TKernel_00024Derived_doMore(JNIEnv* env, jobject obj, jobject arg0) {
    ::Derived* ptr = (::Derived*)jlong_to_ptr(env->GetLongField(obj, JavaCPP_addressFID));
    if (ptr == NULL) {
        env->ThrowNew(JavaCPP_getClass(env, 7), "This pointer address is NULL.");
        return 0;
    }
    jlong position = env->GetLongField(obj, JavaCPP_positionFID);
    ptr += position;
    ::Base* ptr0 = arg0 == NULL ? NULL : (::Base*)jlong_to_ptr(env->GetLongField(arg0, JavaCPP_addressFID));
    jlong size0 = arg0 == NULL ? 0 : env->GetLongField(arg0, JavaCPP_limitFID);
    void* owner0 = JavaCPP_getPointerOwner(env, arg0);
    jlong position0 = arg0 == NULL ? 0 : env->GetLongField(arg0, JavaCPP_positionFID);
    ptr0 += position0;
    size0 -= position0;
    MySmartPtrAdapter< ::Base > adapter0(ptr0, size0, owner0); <<<<<<<<<<<<<<< LINE 1
    jobject rarg = NULL;
    ::Base* rptr;
    jthrowable exc = NULL;
    try {
        MySmartPtrAdapter< ::Base > radapter(ptr->doMore(adapter0)); <<<<<<<<<<<<<<< LINE 2
        rptr = radapter;
        jlong rcapacity = (jlong)radapter.size;
        void* rowner = radapter.owner;
        void (*deallocator)(void*) = rowner != NULL ? &MySmartPtrAdapter< ::Base >::deallocate : 0;
        if (rptr != NULL) {
            rarg = JavaCPP_createPointer(env, 8);
            if (rarg != NULL) {
                JavaCPP_initPointer(env, rarg, rptr, rcapacity, rowner, deallocator);
            }
        }
    } catch (...) {
        exc = JavaCPP_handleException(env, 9);
    }

    ::Base* rptr0 = adapter0;
    jlong rsize0 = (jlong)adapter0.size;
    void* rowner0 = adapter0.owner;
    if (rptr0 != ptr0) {
        JavaCPP_initPointer(env, arg0, rptr0, rsize0, rowner0, &MySmartPtrAdapter< ::Base >::deallocate);
    } else {
        env->SetLongField(arg0, JavaCPP_limitFID, rsize0 + position0);
    }
    if (exc != NULL) {
        env->Throw(exc);
    }
    return rarg;
}
```

In above an instance of `MySmartPtrAdapter` is created twice on the lines I've marked with 1 and 2.

LINE 1 is as it should be AFAIU.

LINE 2 is not what I want, I think.

Either I would like to invoke some other adapter class there that would create the correct Java class (`Derived` in the examle) based  on the RTTI name I get with `DynamicType()` or (preferably) the `doMore()` on the Java side would not be native, but would instead woke some adapter function that would call a native function to obtain the class C++ pointer value and from this invoke the ’DynamicType` and create a Java class based on that name and the pointer value and return that.

I'm visioning something like this on the generated Java code:

```
 public native @MySmartPtr Base __doMore(@MySmartPtr Base p);

 public Base doMore(@MySmartPtr Base p) {
    Base t =   __doMore(p	);
    String cname=t.DynamicType();
    clazz = Class.forName(cname);
    Constructor ct = clazz.getDeclaredConstructor(Long.TYPE, Boolean.TYPE);
    return ct.newInstance(t.getPointer(), true);
    }
```

Or any other way to get the doMore() on the Java side to return the correct Java class based on the dynamic type of the object returned from the native library.

**Reply by saudet** [reply]
Like I said, you won't be able to do anything about that in C++, with the current state of JavaCPP. It needs to be enhanced. However, you should be able to do everything you need to do in Java. Which part of this are you not able to figure out how to do from Java?

**Reply by nyholku** [reply]
Referring to my example above:

1. How do I cause a C++ function named `FOO` that returns `opencascade::hande` to be renamed `_FOO` and the corresponding JNI C++ and Java declarations/definitions to be generated.

2. How do I cause a Java function called `FOO` to be created that will call `_FOO` and do the new instance stuff as explained above.

Note that there are literally thousands of such functions in the OCC library so I cannot possibly list them by name, I need to have something that does above based on the fact that the function in question is returning an `opencascade::hande`.

**Reply by saudet** [reply]
Like I keep telling you, that can't be done with the current state of JavaCPP. You'll need to do the downcasting manually from Java, provide some helper methods, etc. Unless of course you would like to make a contribution and make that work nicely, that would be great! You cannot expect me to do everything for you, unless there is money involved, obviously.

**Reply by nyholku** [reply]
Sorry I did not understood that. 

I was convinced that there is such mechanism and that I was just too dump to figure it out. I am not expecting anyone to do anything, I was just looking for information how to do this because I was under the misapprehension that of course this sort of basic functionality would be there. 

My bad.

I will have a look to see if this is something I could contribute. 

As to money, we are both in the same boat, I'm doing this as labor of love to heltp to  contribute modern Java bindings for OCC which is an amazing toolkit collection for doing CSG.


Thank you for your attention and time.

**Reply by saudet** [reply]
> I was convinced that there is such mechanism and that I was just too dump to figure it out. I am not expecting anyone to do anything, I was just looking for information how to do this because I was under the misapprehension that of course this sort of basic functionality would be there.

It's not "basic functionality", none of the libraries listed here, except Qt, needs RTTI: https://github.com/bytedeco/javacpp-presets

And I keep telling you, we can very easily provide a couple of helper methods. Your `downcastImpl()` works pretty much as is with JavaCPP. Why is this unsatisfactory? Why do you need perfection? Unless you are willing to spend extraordinary amounts of energy and time in this, no solutions is every going to be perfect! Please be realistic in your expectations.

**Reply by nyholku** [reply]
I don't think we are communicating very effective and productively. 

You keep saying that you keep telling me and I don't see what you are telling me.

For instance, what are those "easily provide a couple of helper methods" where are they?

Anyways, I'm not looking for perfection. But I do think that I want the users of OCC be able to do:

```
Geom_Gemoetry obj = SomeToolkit.someFunction();
if (obj instanceof Geom_Plane) {
   Geom_Plane asPlane = (Geom_Plane) obj;
   asPlane.foobar_from_Geom_Plane_Class();
```

instead of something like this:

```
Geom_Gemoetry obj = SomeToolkit.someFunction();
if (obj.DynamicType().name.equals("Geom_Plane))
     Geom_Plane geomObj = new Geom_Plane(obj.getPointer());
     asPlane.foobar_from_Geom_Plane_Class();
```

**Reply by saudet** [reply]
With your helper method `downcastImpl()` you could do something like `downcastImpl(obj) instanceof Geom_Plane` or put it as a non-static method like `obj.downcastImpl() instanceof Geom_Plane`, but you're saying that's not good enough, so, yes, you are seeking perfection. That's fine, but you need to keep your expectations in line with your resources, so to get perfection you'll need to increase the resources that you are willing to allocate into solving this problem. I also already told you that contributions are welcome. If you'd like to do that, I will be happy to assist you with issues that you may encounter in doing so, but you will need to show a little bit of effort on your part.

**Comment by nyholku** [comment]
Hi,

I took a stab at providing a new feature for JavaCPP that allows adding automatic downcasting if so desired.

This is very much WIP and a first stab and I could envision a lot of different ways to do that.

You can find the code here:

https://github.com/nyholku/javacpp/tree/feature/autodowncast

It is very simple.

I added a parameter `downcast` to `@Adapter`. The default value is false so it is downwards compatible (I hope) with any existing annotations.

Then I modified the code generation to generate a call to ` jobject createPointer(JNIEnv* env)` in the adapter template class if the `downcast` value is true (instead of calling the `JavaCPP_createPointer`). If the parameter is false then the generated code is a before. 

The user then needs to implement this method in the adapterd and do what ever it takes to dynamically downcast the pointer that the adapter hosts.

I'm in no way sure memory is managed correctly, but a trivial test case runs.

For OCC I've create an adapter template class, here:

https://github.com/nyholku/javacpp-occ/blob/master

specifically here:

https://github.com/nyholku/javacpp-occ/blob/master/src/occ-handle-adapter.hxx

There is a fake C++ OCC API included that uses the OCC class and calling convention here:

https://github.com/nyholku/javacpp-occ/blob/master/src/FakeOccDemoApi.hxx

and a Java demo that demonstrates calling an API function that returns calls  with OCC handle to the native API and returns a downcast correct Java class from that call more or less automatically.

https://github.com/nyholku/javacpp-occ/blob/master/src/OccDowncastDemo.java

The `build.sh` is all over the place and only works on my machine without mods and it requires the OCC headers.

This is in no way representative how the OCC Java bindings should or perhaps will be done and structured, but the https://github.com/nyholku/javacpp-occ/blob/master/src/occ/TKernelConfig.java demonstrates how some of the fundamental OCC headers can be used as-is with my modified JavaCPP.

All comments are welcome.


(Also fixed an array subscript error in Parser IIRC.)

**Reply by saudet** [reply]
That's a step in the right direction, thanks! Please open a draft pull request with this instead of posting a new "answer" here though.

However, with your proposal, users need to code in JNI, and that's something I want to avoid. Since JNI is just a bad way of coding what should be done in Java anyway, what about letting the user write a method in Java, and use an annotation maybe like `@Downcaster Base downcast() { /* custom user code in Java */ }` that tells JavaCPP to call that method before returning a new object?
